scanpy/experimental/pp/_highly_variable_genes.py

```diff
@@ -0,0 +1,341 @@
+
+    view_to_actual(adata)
+    X = _get_obs_rep(adata, layer=layer)
+    computed_on = layer if layer else 'adata.X'
+
+    # Check for raw counts
+    if check_values and (check_nonnegative_integers(X) is False):
+        warnings.warn(
+            "`flavor='pearson_residuals'` expects raw count data, but non-integers were found.",
+            UserWarning,
+        )
+    # check theta
+    if theta <= 0:
+        # TODO: would "underdispersion" with negative theta make sense?
+        # then only theta=0 were undefined..
+        raise ValueError('Pearson residuals require theta > 0')
+    # prepare clipping
+
+    if batch_key is None:
+        batch_info = np.zeros(adata.shape[0], dtype=int)
+    else:
+        batch_info = adata.obs[batch_key].values
+    n_batches = len(np.unique(batch_info))
+
+    # Get pearson residuals for each batch separately
+    residual_gene_vars = []
+    for batch in np.unique(batch_info):
+
+        adata_subset = adata[batch_info == batch]
+
+        # Filter out zero genes
+        with settings.verbosity.override(Verbosity.error):
+            nonzero_genes = filter_genes(adata_subset, min_cells=1, inplace=False)[0]
+        adata_subset = adata_subset[:, nonzero_genes]
+
+        if layer is not None:
+            X_batch = adata_subset.layers[layer]
+        else:
+            X_batch = adata_subset.X
+
+        # Prepare clipping
+        if clip is None:
+            n = X_batch.shape[0]
+            clip = np.sqrt(n)
+        if clip < 0:
+            raise ValueError("Pearson residuals require `clip>=0` or `clip=None`.")
+
+        if sp_sparse.issparse(X_batch):
+            sums_genes = np.sum(X_batch, axis=0)
+            sums_cells = np.sum(X_batch, axis=1)
+            sum_total = np.sum(sums_genes).squeeze()
+        else:
+            sums_genes = np.sum(X_batch, axis=0, keepdims=True)
+            sums_cells = np.sum(X_batch, axis=1, keepdims=True)
+            sum_total = np.sum(sums_genes)
+
+        # Compute pearson residuals in chunks
+        residual_gene_var = np.empty((X_batch.shape[1]))
+        for start in np.arange(0, X_batch.shape[1], chunksize):
+            stop = start + chunksize
+            mu = np.array(sums_cells @ sums_genes[:, start:stop] / sum_total)
+            X_dense = X_batch[:, start:stop].toarray()
+            residuals = (X_dense - mu) / np.sqrt(mu + mu ** 2 / theta)
+            residuals = np.clip(residuals, a_min=-clip, a_max=clip)
+            residual_gene_var[start:stop] = np.var(residuals, axis=0)
+
+        # Add 0 values for genes that were filtered out
+        zero_gene_var = np.zeros(np.sum(~nonzero_genes))
+        residual_gene_var = np.concatenate((residual_gene_var, zero_gene_var))
+        # Order as before filtering
+        idxs = np.concatenate((np.where(nonzero_genes)[0], np.where(~nonzero_genes)[0]))
+        residual_gene_var = residual_gene_var[np.argsort(idxs)]
+        residual_gene_vars.append(residual_gene_var.reshape(1, -1))
```

Comment:
Should this be equivalent? It'd be a bit shorter/ have less logic to follow.

```suggestion
        unmasked_residual_gene_var = np.zeros(len(nonzero_genes))
        unmasked_residual_gene_var[nonzero_genes] = residual_gene_var
        residual_gene_vars.append(unmasked_residual_gene_var.reshape(1, -1))
```

Reply:
Yes, very good! Changed as suggested.